### PR TITLE
planner: initial work for planner

### PR DESCRIPTION
**Pull request by dongahn**
Planner provides a simple abstraction and efficient mechanisms to keep track of the resource states for batch-job scheduling.

- `resource/planner/README.md` has a high-level description on the API and its mechanisms which are based on top of Linux OS kernel's red-black binary search tree code.
    
- `planner.h` has in-lined documentation for its API.
    
- Examples can be found in a unit test: `resource/planner/test/planner_test01.cpp`

Add C++ support into our build system in anticipation for future C++-based development.

### CODE REVIEW

**Comment by dongahn** [comment]
Travis caught an issue with a Makefile. I am on it.

**Comment by dongahn** [comment]
Sorry for all the noise. This is now ready for your review. Thanks!

File: resource/planner/test/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,23 @@
+AM_CXXFLAGS = \
+	@GCCWARN@ \
```
Comment (morrone) [inline resource/planner/test/Makefile.am]: The @GCCWARN@ here should be $(WARNING_CXXFLAGS), but that variable needs fixing.

File: configure.ac [hunk 2]
```diff
@@ -18,10 +18,18 @@ LT_INIT([dlopen])
...
 AC_PROG_LN_S
 AC_PROG_MAKE_SET
 AM_PROG_CC_C_O
+
+AC_PROG_CXX
+AX_CXX_COMPILE_STDCXX([11], [noext], [mandatory])
+if test "$GXX" = "yes"; then
+  WARNING_CXXFLAGS=$WARNING_CFLAGS
+  AC_SUBST([WARNING_CXXFLAGS])
+fi
```
Comment (morrone) [inline configure.ac]: This hunk should probably be in the commit that added the AX_CXX_COMPILE_STDCXX rather than in the planner commit.  Not a big deal though.

File: resource/planner/planner.c [hunk 3]
```diff
@@ -0,0 +1,1271 @@
...
+#include "src/common/libutil/rbtree.h"
+#include "src/common/libutil/rbtree_augmented.h"
+#include "src/common/libutil/xzmalloc.h"
+#include "planner.h"
+
+#define START(node) ((node)->start)
+#define LAST(node)  ((node)->last)
+
+typedef struct rb_root rb_root_t;
+typedef struct rb_node rb_node_t;
```
Comment (morrone) [inline resource/planner/planner.c]: I'm really not a fan of the above two local aliases.  Since the library doesn't provide them, I think it would be best not to use aliases here.

File: resource/planner/planner.c [hunk 4]
```diff
@@ -0,0 +1,1271 @@
...
+        } else {
+            return this_data;
+        }
+    }
+    return last_state;
+}
+
+static int scheduled_point_insert (scheduled_point_t *new_data, rb_root_t *root)
+{
+    rb_node_t **link = &(root->rb_node);
```
Comment (morrone) [inline resource/planner/planner.c]: Wouldn't this be equivalent, and much more readable, if this was an rb_node_t * instead of an rb_note_t **?

As far as I can tell, "link" is only a ** so that it can be used in rb_link_node().  But that is just the _return value_ from rb_link_node(), which we are throwing away.  So that usage of "link" is unrelated to the usages of "link" earlier in the function.  You might want to add a separate pointer for rb_link_node's return pointer just to make that distinction clear.

It looks like there are other places in this file where the same thing may be happening.

Comment (morrone) [reply resource/planner/planner.c]: Errr, nevermind.  I think I see what is going on now. :)

Comment (dongahn) [reply resource/planner/planner.c]: Yeah... I am not a big fan of the interface part of the Linux kernel RB tree. Brian B's perspective was "this is designed for speed; not for interface". FWIW, https://lwn.net/Articles/184495/ was pretty helpful.

File: configure.ac (hunk 2 above)
Comment (dongahn) [reply configure.ac]: Yes, I was going to add this in.

File: resource/planner/planner.c (hunk 3 above)
Comment (dongahn) [reply resource/planner/planner.c]: :-) One of the reasons for `typedef` for me is reduce typing. In particular w/ C++'s very long template types, use of `typedef` for this purpose is essential. But this is C code and if you think this will increase readability, I am more than happy to use the original types.

**Comment by dongahn** [comment]
> My first request would be to add a reference to red/black trees to make it easier for the reader to understand the foundation of what you're building on. I suggest http://pages.cs.wisc.edu/~paton/readings/Red-Black-Trees/ or https://en.wikipedia.org/wiki/Red%E2%80%93black_tree - whichever you prefer.

Good point. I will add a reference for RB Tree and then also the kernel article about its implementation/API.

**Comment by dongahn** [comment]
> There are two red/black trees attached to every resource. For those resources that are part of a dominant hierarchy, there are two red/black trees assigned to every resource in the hierarchy.

Yes, and these search trees will be hidden under a planner object. (You will see from my upcoming `resource-query` work.)

File: resource/planner/planner.c (hunk 3 above)
Comment (morrone) [reply resource/planner/planner.c]: Yes, I think readability is more important that reduced typing.  The kernel guys don't allow wrapping every struct in a typedef, and I think there is good reason not to do that.  I think the typedef just adds a layer of indirection for the reader.  If the typedef and the struct declaration are in the same place it isn't quite as bad, but when the alias is completely local to one file and separate from the struct declaration, I think that makes things harder to follow.

I think in C, typedefs should be restricted to things like making an opaque handle.

**Comment by dongahn** [comment]
> The second r/b tree assigned to each resource contains objects that contain minimum time. For example the above resource might have the first object of this min-time r/b tree be time c. (and here is where I need clarification.)

Mintime resource tree is much more general than this. Using your example,

> Therefore, each resource is assigned one red/black tree that tracks the allocation of size units over time. For example, the first r/b tree might have objects that contain 5 allocated units from time a to time b, 3 units from time b to c, and zero allocated units from time c to infinity.

Say, the resource has 10 unit total and a request wants to find about the minimum time where it can get 7 units. Then, the mintime search tree will answer `b` being the minimum time because obvious you can't get 7 units at the time a. On the other hand, if you want to find about the minimum time where it can get 5 units, it will answer `a`. 

The mistime tree answers each query in O(log N).

File: resource/planner/planner.c (hunk 3 above)
Comment (dongahn) [reply resource/planner/planner.c]: good to know. thanks.

**Comment by dongahn** [comment]
Mintime tree is essential for reservation operation as we don't really want to do full tree search for each and all of the resource changing state points... Essentially, what the planner allows me to do is to prune both time and resource graph searches...

**Comment by dongahn** [comment]
> So, perhaps you can see my confusion and insert text that clarifies the design description.

Yes, I will add this. In general, given the complexity of all of these, my strategy is to do a conference paper submission to describe all of the design decisions, performance tradeoffs etc.

**Comment by dongahn** [comment]
> Also a question: will this design accommodate a resource allocation that is not monotonically decreasing? For example, in my above example, would the search algorithm support a case where there was a "d" object in the scheduled point tree which starts in 2 days and spans 5 minutes? IOW, would the search avoid this resource if it is being considered for a job that requests the resource for an entire week? The resource is available after object c, but then in 2 days, it's allocated again.

Yes, as you can see from some of the APIs, "availability" check includes the span check. If the count requirement cannot satisfied for the entire span, the resource vertex isn't available.

File: resource/planner/planner.h [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+typedef struct span span_t;
+typedef struct planner planner_t;
+
+/*! Construct a planner.
+ *
+ *  \param base_time    earliest schedulable point expressed in integer time
+ *                      (i.e., the base time of the planner to be constructed).
+ *  \param duration     time span of this planner (i.e., all planned spans
+ *                      must end before base_time + duration).
+ *  \param resource_totals
```
Comment (lipari) [inline resource/planner/planner.h]: is resource_totals the same as resource_counts?

Comment (dongahn) [reply resource/planner/planner.h]: Yes, and good catch. I will change the `resource_counts` in the comment to `resource_totals`.

**Comment by dongahn** [comment]
FYI -- I rebased this PR to PR #270 level. But I won't force a push yet since this can confuse reviewers. I will wait until at least one of the reviewers (e.g., @lipari) complete the review and incorporate all the changes requested by then into that push.

File: resource/planner/planner.h (hunk 5 above)
Comment (dongahn) [reply resource/planner/planner.h]: @lipari: Looking through the comments; I found other issues (typos and out-of-sync documents) and fixed some of these in my PR branch. Feel free to surface up issues in the comments as you will surely find. Thanks.

File: resource/planner/planner.c [hunk 6]
```diff
@@ -0,0 +1,1271 @@
...
+static int64_t mintime_resource_subtree_min (scheduled_point_t *point)
+{
+    int64_t min = point->at;
+    scheduled_point_t *p = NULL;
+    if (point->resource_rb.rb_left) {
+        p = rb_entry (point->resource_rb.rb_left, scheduled_point_t, resource_rb);
+        if (min > p->subtree_min)
+            min = p->subtree_min;
+    }
+    if (point->resource_rb.rb_right) {
```
Comment (lipari) [inline resource/planner/planner.c]: Without really understanding what is going here on (yet), the logic looks odd.  I would expect an "else if" here.  As written, the min variable can get assigned in the clause above and then overwritten in this clause.  Is that intended?

Comment (dongahn) [reply resource/planner/planner.c]: This can be confusing, but this is correct code. 

Essentially, how linux augmented rbtree interface works is that you have to provide a callback function which allows the tree to maintain an accurate subtree quantify (in my case, the minimum time of the subtree). So, this function checks both branches and return the min between two.

What min-time resource tree does *essentially* is

1) descend to a node where a resource count can be satisfied

2) and then use the minimum time of its subtree to find the schedule point that has the minimum time. 

There are some corner cases but this is the main logic.

In a sense, it is similar to how you can use the augmented rbtree to implement an interval tree. Perhaps, this article can help you to understand the log behind this a bit better (look for interval tree sample usage):

https://www.kernel.org/doc/Documentation/rbtree.txt

**Comment by lipari** [comment]
> FYI -- I rebased this PR to PR #270 level. But I won't force a push yet since this can confuse reviewers. I will wait until at least one of the reviewers (e.g., @lipari) complete the review and incorporate all the changes requested by then into that push.

Go ahead and force a push of your latest changes, @dongahn.  I'd like another day to complete the review, but at this point, it would help to review all the latest changes.

**Comment by dongahn** [comment]
@lipari: okay, pushed.

**Comment by dongahn** [comment]
FYI -- today's performance study revealed that there is a bug in the scheduled point reference counting scheme. I will patch this one liner next time I will force a push and add more "span remove" unit tests.

```diff
@@ -1066,6 +1064,7 @@ int64_t planner_add_span (planner_t *ctx, int64_t start_time,
     start_point->new_point = 0;
     span->start_p = start_point;
     last_point->new_point = 0;
+    last_point->ref_count++;
     span->last_p = last_point;

     update_mintime_resource_tree (ctx, list);

```

File: resource/planner/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+computing (HPC) is undergoing significant changes
+in both performance-limiting resource types
+and workloads, simple data models
+(e.g., bitmaps) representing HPC compute and other
+consumable resources present increasingly greater
+challenges for effective scheduling. In response,
+Flux is in the process of modeling complex HPC
+resources using a graph. While this
+allows a highly sophisticated resource selections,
+it can be done at the exchange of excessive, highly
```
Comment (lipari) [inline resource/planner/README.md]: shouldn't this read:  "it can be done at the _expense_ of excessive..."

Comment (dongahn) [reply resource/planner/README.md]: That'd be better. I will change it.

File: resource/planner/planner.h [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+#endif
+
+#define PLANNER_NUM_TYPES 5
+
+typedef struct span span_t;
+typedef struct planner planner_t;
+
+/*! Construct a planner.
+ *
+ *  \param base_time    earliest schedulable point expressed in integer time
```
Comment (lipari) [inline resource/planner/planner.h]: "integer time" seems overly non-committal.  I suspect that this will default to seconds.  But then, I wonder is it epoch seconds?  Unless you see a scenario where integer time represented something else (e.g., milliseconds), would it be asking too much to commit to "seconds".  And if you do that, you may want to qualify that as epoch seconds.  Why?  It helps to remove a layer of abstraction which is (I propose) unnecessary.

File: resource/planner/planner.h [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+/*! Find and return the earliest point in integer time when the request can be
+ *  satisfied.
+ *
+ *  Note on semantics: this function returns a time point where the resource state
+ *  changes. If the number of available resources change at time point t1 and
+ *  t2 (assuming t2 is greater than t1+2), the possible schedule points that this
+ *  function can return is either t1 or t2, but not t1+1 nor t1+2 even if these
+ *  points also satisfy the request.
+ *
```
Comment (lipari) [inline resource/planner/planner.h]: > but not t1+1 nor t1+2 even if these points also satisfy the request.

I think this is because, in this example, there is no resource state change in t1+1 or t1+2.  If that's true, you might just drop the discussion of t1+1 and t1+2.  Or perhaps I am missing some relevance to the t1+1 and t1+2 times.

**Comment by lipari** [comment]
I've finished my review and recommend merging pending the reviews of the other reviewers.  My only thought comes in the use of this new code.  As I see it, the span objects are generic - they can be used to make allocations or reservations.  Operationally, once the planner code is put to use, the spans representing the allocations will remain resident in the planner object over time - and removed once the job terminates.  The reservations will come and go with each scheduling loop.  I didn't see a mechanism that will allow you to handle allocations different from reservations.  And I wonder whether such a mechanism to lock in an allocation span should be part of the planner rudiments.

File: resource/planner/planner.h (hunk 8 above)
Comment (dongahn) [reply resource/planner/planner.h]: >  I suspect that this will default to seconds.

While I agree our upper level software (scheduling infrastructure) will program this to seconds, this really depends on the users of the planner API. 

If one wants to introduce a coarser grained scheduling, a tick could be programmed to be coarser than a second; if fine-grained scheduling, the unit could be finer than a second. 

Point is, the planner doesn't want to care about these physical time units and want to remain to treat time as unit-less. 1 just represents 1 tick in the upper level software's time granularity.

I believe this is a better separation of concerns.

Comment (lipari) [reply resource/planner/planner.h]: That's what I suspected.  No problem.

File: resource/planner/planner.h (hunk 9 above)
Comment (dongahn) [reply resource/planner/planner.h]: This is just giving an example to make the previous sentence a bit more concrete. Neither t1+1 nor t1+2 has a resource state change. Unless this confuses folks too much, I think it makes sense to keep it.

**Comment by dongahn** [comment]
Thank you for your thorough review @lipari!

> I didn't see a mechanism that will allow you to handle allocations different from reservations. And I wonder whether such a mechanism to lock in an allocation span should be part of the planner rudiments.

Good thought. But I need a bit more progress to further determine whether I need other idioms.

**Comment by dongahn** [comment]
OK. I have made all the changes requested so far including @morrone's avoiding typedefs and forced a push.

File: src/common/libutil/Makefile.am [hunk 10]
```diff
@@ -16,4 +16,8 @@ libutil_la_SOURCES = \
 	log.c \
 	log.h \
 	oom.h \
-	shortjansson.h
+	shortjansson.h \
+	compiler.h \
+	rbtree.c \
```
Comment (morrone) [inline src/common/libutil/Makefile.am]: I think it would probably be best to put rbtree into its own helper library rather than mixing with libutil.  rbtree is third-party GPL code, so anything that uses it will be ineligible for inclusion in LGPL libraries that we might want to export.  Keeping it separate will make it easier to avoid accidental inclusion in the wrong place.

**Comment by dongahn** [comment]
Thanks @morrone for your thorough review! I skimmed through them, and I believe most, if not all, of your suggestions will improve the quality of the planner. I have a visitor tomorrow so i will try to address them this Friday.

File: resource/Makefile.am [hunk 11]
```diff
@@ -0,0 +1,10 @@
+AM_CFLAGS = \
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/Makefile.am]: Make sense.

File: resource/planner/planner.c [hunk 12]
```diff
@@ -0,0 +1,1275 @@
...
+{
+    int i = 0;
+    ctx->plan_start = base_time;
+    ctx->plan_end = base_time + (int64_t)duration;
+    ctx->sched_point_tree = RB_ROOT;
+    ctx->mt_resource_tree = RB_ROOT;
+    ctx->p0 = xzmalloc (sizeof (*(ctx->p0)));
+    ctx->p0->at = base_time;
+    ctx->p0->ref_count = 1;
+    memset (ctx->p0->scheduled, '\0', sizeof (ctx->p0->scheduled));
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/planner.c]: I didn't realize xzmalloc actually zeroes the buffer. Thanks.

File: resource/planner/planner.c [hunk 13]
```diff
@@ -0,0 +1,1275 @@
...
+}
+
+int planner_reset (planner_t *ctx, int64_t base_time, uint64_t duration)
+{
+    if (duration < 1) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    erase (ctx);
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/planner.c]: Wow. Great catch, @morrone!

File: resource/planner/planner.c [hunk 14]
```diff
@@ -0,0 +1,1275 @@
...
+    initialize (ctx, base_time, duration);
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/planner.c]: This initial pseudo schedule point should be freed by `scheduled_points_destroy()`

File: resource/planner/planner.h [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+ *                      are supported.
+ *  \param resource_types
+ *                      string array of size len where each element contains
+ *                      the resource type corresponding to each corresponding
+ *                      element in the resource_totals array.
+ *  \param len          length of the resource_totals and resource_types arrays;
+ *                      must not exceed PLANNER_NUM_TYPES.
+ *  \return             new planner context; NULL on an error with errno set
+ *                      as follows:
+ *                      0 on success; -1 on an error with errno set:
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/planner.h]: Sure.

File: resource/planner/planner.h [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+ */
+int planner_reset (planner_t *ctx, int64_t base_time, uint64_t duration);
+
+/*! Destroy the planner.
+ *
+ *  \param ctx_p        pointer to a planner context pointer returned.
+ *                      from planner_new
+ *
+ */
+void planner_destroy (planner_t **ctx_p);
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/planner.h]: For now, I will just go with as-is as this is a common form of flux interfaces. If we later decide the other form has a clear advantage, we can make this change globally.

File: src/common/libutil/Makefile.am (hunk 10 above)
Comment (dongahn) [reply src/common/libutil/Makefile.am]: Great suggestion. This makes perfect sense!

File: resource/planner/planner.h [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+ *                      the requested resource counts.
+ *  \param len          length of resource_counts and resource_types arrays.
+ *                      must not exceed PLANNER_NUM_TYPES.
+ *  \return             earliest time at which the request can be satisfied;
+ *                      -1 on an error with errno set as follows:
+ *                          EINVAL: invalid argument.
+ *                          ERANGE: resource_counts contain an out-of-range value.
+ *                          ENOTSUP: internal error encountered.
+ */
+int planner_avail_during (planner_t *ctx, int64_t at, uint64_t duration,
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/planner.h]: Ah. I can see why this could be confusing with "at" in the interface. I will go with your suggestion.

File: resource/planner/README.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+Flux is in the process of modeling complex HPC
+resources using a graph. While this
+allows a highly sophisticated resource selections,
+it can be done at the expense of excessive, highly
+time-consuming searches of a large graph.
+
+Planner's efficient mechanism is specifically designed
+to mitigate the scalability and performance risk
+of such a graph-based scheduling approach. More specifically,
+it enables a novel scheme called scheduler-driven
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/planner/README.md]: I think the "scheduler-driven aggregate updates" scheme would need a more in-depth and formal technical writing (probably a section in a conference paper later). I will just genericize and shorten this to one sentence.

**Comment by dongahn** [comment]
Given all the structural changes I have, it makes more sense to do a fresh PR. Closing this and will post a new PR shortly.
